src/v/kafka/server/usage_manager.cc

```diff
@@ -0,0 +1,381 @@
+      "Error correctly detecting last window delta");
+    _timer.arm((now + _usage_window_width_interval) - last_window_delta);
+    _persist_disk_timer.arm(now + _usage_disk_persistance_interval);
+}
+
+std::vector<usage_window>
+usage_manager::accounting_fiber::get_usage_stats() const {
+    std::vector<usage_window> stats;
+    for (size_t i = 1; i < _buckets.size(); ++i) {
+        const auto idx = (_current_window + i) % _usage_num_windows;
+        if (!_buckets[idx].is_uninitialized()) {
+            stats.push_back(_buckets[idx]);
+        }
+    }
+    /// Open bucket last ensures ordering from oldest to newest
+    stats.push_back(_buckets[_current_window]);
+    /// std::reverse returns results in ordering from newest to oldest
+    std::reverse(stats.begin(), stats.end());
+    return stats;
+}
+
+ss::future<> usage_manager::accounting_fiber::stop() {
+    _timer.cancel();
+    _persist_disk_timer.cancel();
+    try {
+        co_await persist_to_disk(
+          _kvstore,
+          persisted_state{
+            .configured_period = _usage_window_width_interval,
+            .configured_windows = _usage_num_windows,
+            .current_state = _buckets.copy()});
+    } catch (const std::exception& ex) {
+        vlog(
+          klog.debug,
+          "Encountered exception when persisting usage data to disk: {}",
+          ex);
+    }
+    co_await _gate.close();
+}
+
+ss::future<> usage_manager::accounting_fiber::close_window() {
+    const auto now = ss::lowres_system_clock::now();
+    _buckets[_current_window].u = co_await _um.map_reduce0(
+      [](usage_manager& um) { return um.sample(); },
+      _buckets[_current_window].u,
+      [](const usage& acc, const usage& x) { return acc + x; });
+    _buckets[_current_window].end = epoch_time_secs(now);
+    _current_window = (_current_window + 1) % _buckets.size();
+    _buckets[_current_window].reset(now);
+}
+
+std::chrono::seconds usage_manager::accounting_fiber::reset_state(
+  fragmented_vector<usage_window> buckets) {
+    /// called after restart to determine which bucket is the 'current' bucket
+    auto last_window_delta = std::chrono::seconds(0);
+    _current_window = 0;
+    if (!buckets.empty()) {
+        std::optional<size_t> open_index;
+        for (size_t i = 0; i < buckets.size(); ++i) {
+            /// There will always be exactly 1 open_window in the result set
+            if (buckets[i].is_open()) {
+                vassert(!open_index, "Data serialization was incorrect");
+                open_index = i;
+                break;
+            }
+        }
+        vassert(open_index, "Data serialization was incorrect");
+        /// Optimization to begin picking up if wall time is within
+        /// window interval
+        const auto begin = std::chrono::seconds(buckets[*open_index].begin);
+        const auto now = std::chrono::seconds(epoch_time_secs());
+        vassert(begin <= now, "Clock or data serialization error");
```

Comment:
Forgot to reply the your first comment. I chose to go with a system clock because I was getting incorrect results when converting the low_res clock to seconds since epoch

Reply:
nothing wrong with using the system clock. i'm asking: because you are using the system clock, does this assertion hold?

Reply:
are you suggesting to remove the vassert because clock drift could result in a situation where the assertion fails to hold?

Reply:
I see your point, the min window width is 1s, so the drift would have to be at least 1 second behind.

Reply:
i think a good solution would be to just remove the vassert and let the program open a new window

Reply:
Resolved by just continuing at the last open window if this condition is detected

Reply:
i'm not even talking about drift. i'm saying that someone (e.g. ntpd) could just change the time on the host at any moment.

Reply:
gotcha, i've added logic to handle it gracefully